src/propagators/zipkin_b3_text_map_codec.js

```diff
@@ -0,0 +1,168 @@
+// @flow
+// Copyright (c) 2017 The Jaeger Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+
+import * as constants from '../constants.js';
+import Metrics from '../metrics/metrics.js';
+import NoopMetricFactory from '../metrics/noop/metric_factory';
+import SpanContext from '../span_context.js';
+import Utils from '../util.js';
+import { parseCommaSeparatedBaggage } from '../propagators/baggage';
+
+const ZIPKIN_PARENTSPAN_HEADER = 'x-b3-parentspanid';
+const ZIPKIN_SPAN_HEADER = 'x-b3-spanid';
+const ZIPKIN_TRACE_HEADER = 'x-b3-traceid';
+const ZIPKIN_SAMPLED_HEADER = 'x-b3-sampled';
+const ZIPKIN_FLAGS_HEADER = 'x-b3-flags';
+
+export default class ZipkinB3TextMapCodec {
+    _urlEncoding: boolean;
+    _baggagePrefix: string;
+    _metrics: any;
+
+    constructor(options: any = {}) {
+        this._urlEncoding = !!options.urlEncoding;
+        this._baggagePrefix = options.baggagePrefix || constants.TRACER_BAGGAGE_HEADER_PREFIX;
+        this._baggagePrefix = this._baggagePrefix.toLowerCase();
+        this._metrics = options.metrics || new Metrics(new NoopMetricFactory());
+    }
+
+    _encodeValue(value: string): string {
+        if (this._urlEncoding) {
+            return encodeURIComponent(value);
+        }
+
+        return value;
+    }
+
+    _decodeValue(value: string): string {
+        // only use url-decoding if there are meta-characters '%'
+        if (this._urlEncoding && value.indexOf('%') > -1) {
+            return this._decodeURIValue(value);
+        }
+
+        return value;
+    }
+
+    _isValidZipkinId(value: string): boolean {
+        // Validates a zipkin trace/spanID by attempting to parse it as a string of hex digits.
+        //
+        // Note: The Number type in JS cannot represent the full range of 64bit
+        // unsigned ints, so using parseInt() on strings representing 64bit hex
+        // numbers only returns an approximation of the actual value.
+        // Fortunately, we do not depend on the returned value, we are only
+        // using it to validate that the string is a valid hex number (which is
+        // faster than doing it manually).  We cannot use
+        // Int64(numberValue).toBuffer() because it throws exceptions on bad
+        // strings.
+        if (!value) {
+            return true
+        }
+
+        let v = parseInt(value, 16);
+        return !isNaN(v) && v.toString(16) === value;
```

Comment:
are you sure about the last `===` term? Given the comment about precision loss I don't think the toString transformation has to return the same original value. For example
```
v = parseInt('ffffffffffffffff', 16);
18446744073709552000
v
18446744073709552000
v.toString(16)
"10000000000000000"
```

Reply:
Sure, I'll remove that.  

I noticed that in `TextMapCodec`, the "validation" of the trace/span IDs is done by just calling `parseInt(foo, 16)` and checking for `NaN`.  In addition to having the limitations mentioned in the code already, this allows values that aren't strictly composed of hex digits to be accepted as valid IDs, since `parseInt` has [the following behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt):

> If parseInt encounters a character that is not a numeral in the specified radix, it ignores it and all succeeding characters and returns the integer value parsed up to that point. 

So, you get the following behavior from `parseInt`:
```
> parseInt('a', 16)
10
> parseInt('azzzzz', 16)
10
```

Reply:
+1. It's unfortunate that parseInt behaves this way, but in the worse case it will treat invalid string as a valid trace id (which is somewhat similar to creating a new random trace ID), whereas doing the reverse `===` comparison will discard many valid IDs.

Reply:
Yes, agreed. If you wanted to be super strict, you'd validate all the trace IDs with a regex for lower hex digits or something, but it's likely not worth the overhead.